---
title: "DNA Methylation Signature of a Lifestyle-based Resilience Index for Cognitive Health"
subtitle: "Check overlap with eQTm results"
author: 
  - Wei Zhang^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    highlight: pygments
    theme: yeti
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

dir.base <- "../.."
dir.data.aux <- file.path(dir.base,"../DATASETS/Aux_Sync") 
dir.results <- file.path(dir.base, "analysis_results")
dir.results.res <- file.path(dir.results, "resilience")
dir.results.combp <- file.path(dir.results, "combp")
dir.supp <- file.path(dir.base, "data/Supp") 
dir.results.overlap <- file.path(dir.results, "overlap")
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

```{r message = F}
library(tidyverse)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
source("../utility/annotation.R")
```

# Load results

```{r message = F, warning = F}
## Load meta results
results <- read_csv(
  file.path(dir.results.res,
            "Resilience_rlm_model_results_bacon.csv")
) %>% dplyr::select(
  cpg, seqnames, start, end, width, Relation_to_Island,
  UCSC_RefGene_Name, UCSC_RefGene_Group, GREAT_annotation, Estimate.bacon, pValue.bacon
)

## Sig single cpgs
cpgs.sig <- results %>% filter(pValue.bacon < 1e-05) 

## Load comb-p DMR results
sig.dmr <- read_csv(
  file.path(dir.results.combp, "combp_resilience_rlm_filtered_results_annotated.csv")
) 
sig.dmr <- str_split(sig.dmr$cpgs_in_region, ",") %>% unlist() ## 134
sig.dmr <- results %>% filter(cpg %in% sig.dmr) 
```

# Auxillary function

```{r}
check_overlap <- function(cpg_df, idx, colname_prefix, match_column, return_overlap = F, dir.supp, filename, skip = 1, match = "cpg"){

  # Load supp
  supp <- purrr::map(idx, ~readxl::read_xlsx(
    file.path(dir.supp, filename),
    sheet = .,
    skip = skip
  ))

  plyr::l_ply(
    1:length(supp),
    .fun = function(i){
      if(match == "cpg") {
        overlap <- ifelse(cpg_df$cpg %in% supp[[i]][[match_column]],1,0)
      } 
      if(match == "position") {
        overlap <- ifelse(cpg_df$seqnames %in% supp[[i]][[match_column[1]]] & cpg_df$start %in% supp[[i]][[match_column[2]]],1,0)
      }
      
      cpg_df[[colname_prefix[i]]] <<- overlap
    }
  )
  if(return_overlap) {
    res <- plyr::llply(
      1:length(supp),
      .fun = function(i){
        colnames(supp[[i]])[colnames(supp[[i]]) == match_column] <- "cpg"
        dplyr::inner_join(cpg_df[,! colnames(cpg_df) %in% colname_prefix], supp[[i]])
      }
    )
    names(res) <- colname_prefix
    c(list(overlap = cpg_df), res)
  } else {
    cpg_df
  }
}
```

# Study 1

```{r warning = F}
sig_cpg_df <- check_overlap(
  cpgs.sig, 
  idx = c(2,3,6),
  colname_prefix = c("sig_in_cis_cpgs_FHS", "sig_in_trans_cpgs_FHS", "sig_in_cis_cpgs_meta"),
  match_column = "CpG",
  dir.supp = dir.supp,
  filename = "eQTm_FHS_array_ClinicalEpigenetics_13148_2021_1041_MOESM2_ESM.xlsx",
  return_overlap = T
)
sum(sig_cpg_df$overlap$sig_in_cis_cpgs_FHS) #1
sum(sig_cpg_df$overlap$sig_in_trans_cpgs_FHS) #4
sum(sig_cpg_df$overlap$sig_in_cis_cpgs_meta) 

sig_dmr_df <- check_overlap(
  sig.dmr, 
  idx = c(2,3,6),
  colname_prefix = c("sig_in_cis_cpgs_FHS", "sig_in_trans_cpgs_FHS", "sig_in_cis_cpgs_meta"),
  match_column = "CpG",
  dir.supp = dir.supp,
  filename = "eQTm_FHS_array_ClinicalEpigenetics_13148_2021_1041_MOESM2_ESM.xlsx",
  return_overlap = T
)
sum(sig_dmr_df$overlap$sig_in_cis_cpgs_FHS) #28
sum(sig_dmr_df$overlap$sig_in_trans_cpgs_FHS) #4
sum(sig_dmr_df$overlap$sig_in_cis_cpgs_meta)

writexl::write_xlsx(
  c("Single cpgs" = sig_cpg_df,
    "DMR" = sig_dmr_df),
  file.path(dir.results.overlap, "Overlapped_CpGs_with_eQTm_FHS_array_ClinicalEpigenetic.xlsx")
)
```

# Study 2

```{r warning = F}
sig_cpg_df <- check_overlap(
  cpgs.sig, 
  idx = c(2,3),
  colname_prefix = c("sig_in_cis_cpgs_FHS", "sig_in_trans_cpgs_FHS"),
  match_column = "CpG",
  dir.supp = dir.supp,
  filename = "eQTm_FHS_RNAseq_ScientificReport_41598_2023_39936_MOESM1_ESM.xlsx"
)
sum(sig_cpg_df$sig_in_cis_cpgs_FHS)
sum(sig_cpg_df$sig_in_trans_cpgs_FHS)

sig_dmr_df <- check_overlap(
  sig.dmr, 
  idx = c(2,3),
  colname_prefix = c("sig_in_cis_cpgs_FHS", "sig_in_trans_cpgs_FHS"),
  match_column = "CpG",
  dir.supp = dir.supp,
  filename = "eQTm_FHS_RNAseq_ScientificReport_41598_2023_39936_MOESM1_ESM.xlsx"
)
sum(sig_dmr_df$sig_in_cis_cpgs_FHS)
sum(sig_dmr_df$sig_in_trans_cpgs_FHS)

writexl::write_xlsx(
  list("Single cpgs" = sig_cpg_df,
       "DMR" = sig_dmr_df),
  file.path(dir.results.overlap, "Overlapped_CpGs_with_eQTm_FHS_RNAseq_ScientificReport.xlsx")
)
```

```{r warning = F}
sig_cpg_df <- check_overlap(
  cpg = cpgs.sig, 
  idx = 1,
  colname_prefix = c("DMR"),
  match_column = c("Chromosome", "Start Coordinate (hg38)"),
  dir.supp = dir.supp,
  filename = "alz13514-sup-0005-supmat.xlsx",
  skip = 1,
  match = "position"
)
sum(sig_cpg_df$DMR)
sum(sig_cpg_df$DMR)

sig_dmr_df <- check_overlap(
  cpg = sig.dmr, 
  idx = 1,
  colname_prefix = c("DMR"),
  match_column = c("Chromosome", "Start Coordinate (hg38)"),
  dir.supp = dir.supp,
  filename = "alz13514-sup-0005-supmat.xlsx",
  skip = 1,
  match = "position"
)
sum(sig_dmr_df$DMR)
sum(sig_dmr_df$DMR)
```

```{r eval = F}
writexl::write_xlsx(
  list("Single cpgs" = sig_cpg_df,
       "DMR" = sig_dmr_df),
  file.path(dir.results.overlap, "Overlapped_CpGs_with_DMR_AD.xlsx")
)
```

# Session Information

<details>
  <summary>**Session Info**</summary>
```{r session}
sessionInfo()
```
</details>
